torch/utils/checkpoint.py

```diff
@@ -0,0 +1,114 @@
+        inputs = ctx.saved_tensors
+        inputs_list = detach_variable(inputs)
+        with torch.enable_grad():
+            outputs = ctx.run_function(*inputs_list)
+
+        if torch.is_tensor(outputs):
+            outputs = (outputs,)
+        torch.autograd.backward(outputs, args)
+
+        input_grads = None
+        if isinstance(inputs_list, tuple):
+            input_grads = tuple(inp.grad for inp in inputs_list)
+            return (None,) + input_grads
+        elif torch.is_tensor(inputs_list):
+            input_grads = inputs_list.grad
+            return None, input_grads
+
+
+def checkpoint(run_function, *args):
+    r"""Checkpoint a model or part of the model
+
+    Checkpoint works by trading compute for memory. It can be applied on any
+    part of the model. In the forward pass, the model activations are not
+    stored. The forward pass save the inputs tuple and the run_function
+    parameter. In the backwards pass, the saved inputs and run_function is
+    retreived, and the forward pass is done on the model again (non-volatile
+    this time) since we need to get the activations values for calculating the
+    gradient and then the gradients are calculated.
+    WARNING: checkpointing doesn't work with torch.autograd.grad(), but only with
+    torch.autograd.backward()
+
+    Args:
+        run_function : describes what to run in the forward pass of the model or
+                       part of the model. It should also know how to handle
+                       the inputs passed as the tuple. For example, in LSTM,
+                       user passes (activation, hidden), run_function should
+                       correctly use first input as activation and second input
+                       as hidden
+        args:         tuple containing inputs to the run_function
+
+    Returns:
+        Output of running the run_function on *args
+    """
+    return CheckpointFunction.apply(run_function, *args)
+
+
+def checkpoint_sequential(modules, segments, *inputs):
+    r"""A helper function for checkpointing sequential based models.
+
+    For models that are constructed using sequential, they normally are built
+    using various modules. For such models, given a list of modules it executes
+    sequentially, we can divide the model in various segments and checkpoint
+    the segments. All segments except the last will be run in volatile manner.
+    The inputs of each checkpointed segment will be saved for re-running the
+    segment in the backward pass.
+
+    Args:
+        modules: The sequence of modules (comprising the model) to run in order.
+        segments: Number of times chunks to create in the model
+        inputs: tuple containing the inputs to run_function
+
+    Returns:
+        Output of running the modules on *inputs
+
+    Example:
+        >>> modules = [module for k, module in self._modules.items()][0]
```

Comment:
Same issue as before - can't have `._modules` in an example! Can we just accept an `nn.Sequential` as the first argument, and treat it just like we would a list of modules? That would avoid a lot of awkward unpacking

Reply:
what's the reason for not exposing `._modules` to users? I can just remove this example, we are providing tutorial anyways.

Can you recommend me a way how to get what modules are part of the model? we use this to slice up the model in chunks that user wants.

Reply:
@fmassa In this case we're looking for `.children()`, because we only want immediate children

Reply:
alright, I tried `.children()` and that seems to be fine for a smaller test case. thanks @fmassa and @apaszke :)

Reply:
NB can you please still add this at the top:
```
if isinstance(modules, nn.Sequential):
    modules = list(modules.children())
```